torch/fx/graph_module.py

```diff
@@ -0,0 +1,459 @@
+                body.append(f'{node.name} = {_format_target(root_module, node.target)}\n')
+                continue
+            raise NotImplementedError(f'node: {node.op} {node.target}')
+
+        src = ''.join(body)
+        return src, str(self.result), free_vars
+
+def _format_args(args, kwargs):
+    args_s = ', '.join(repr(a) for a in args)
+    kwargs_s = ', '.join(f'{k} = {repr(v)}' for k, v in kwargs.items())
+    if args_s and kwargs_s:
+        return f'{args_s}, {kwargs_s}'
+    return args_s or kwargs_s
+
+def _format_target(base, target):
+    elems = target.split('.')
+    r = base
+    for e in elems:
+        if not e.isidentifier():
+            r = f'getattr({r}, "{e}")'
+        else:
+            r = f'{r}.{e}'
+    return r
+
+def _run_symbolic_forward(root, is_leaf_module):
+    def _use_parameter(graph, a):
+        if isinstance(a, torch.nn.Parameter):
+            for n, p in root.named_parameters():
+                if a is p:
+                    return graph.get_param(n)
+            raise NameError('parameter is not a member of this module')
+        return NotImplemented
+    graph = Graph(arg_handler=_use_parameter)
+    fn = type(root).forward
+
+    co = fn.__code__
+    total_args = co.co_argcount + co.co_kwonlyargcount
+    names_iter = iter(co.co_varnames)
+    next(names_iter)  # skip self
+    args = [root]
+    args.extend(graph.placeholder(next(names_iter)) for name in range(1, total_args))
+
+    if co.co_kwonlyargcount > 0 or co.co_flags & HAS_VARSTUFF:
+        if co.co_flags & inspect.CO_VARARGS:
+            args.append(graph.placeholder('*' + next(names_iter)))
+        if co.co_flags & inspect.CO_VARKEYWORDS:
+            args.append(graph.placeholder('**' + next(names_iter)))
+        fn = _patch_function(fn, len(args))
+
+    args = tuple(args)
+    orig_call = torch.nn.Module.__call__
+
+    def module_call_wrapper(mod, *args, **kwargs):
+        if not is_leaf_module(mod):
+            return orig_call(mod, *args, **kwargs)
+        else:
+            target = _find_module(root, mod)
+            return graph.call_module(target, args, kwargs)
+    try:
+        torch.nn.Module.__call__ = module_call_wrapper
+        graph.output(fn(*args))
+    finally:
+        torch.nn.Module.__call__ = orig_call
+    return graph
+
+
+class GraphModule(torch.nn.Module):
+    def __new__(cls, *args, **kwargs):
+        # each instance of a graph module needs its own forward method
+        # so create a new singleton class for each instance.
+        # it is a subclass of the user-defined class, the only difference
+        # is an extra layer to install the forward method
+        class GraphModuleImpl(cls):
+            pass
+        return super().__new__(GraphModuleImpl)
+
+    def __init__(self, root, graph=None, is_leaf_module=is_leaf_module):
+        super().__init__()
+        self.root = root
+        self.graph = graph if graph is not None else _run_symbolic_forward(root, is_leaf_module)
```

Comment:
Rather than have `_run_symbolic_forward` be called in the ctor, we should have the api be `torch.fx.symbolic_trace(model) -> GraphModule`. This allows us to put the `_run_symbolic_forward` machinery in its own file, making it clear that `GraphModule` is separable from it. `GraphModule` can then just explictly take a Graph as argument. This makes it clearer how to use in transforms: GraphModule can be constructed given a Graph and an existing root module. `symbolic_trace` can generate a Graph from a module and then create a GraphModule from the parts.